src/tests/async.test.ts

```diff
@@ -12,181 +12,187 @@ describe("async module", () => {
       const asyncSquare = async (a: number): Promise<number> => {
-        return new Promise((res) => res(a * a));
-      };
-      const result = await _.map<number, number>(numbers, asyncSquare);
-      assert.deepEqual(result, [1, 4, 9, 16]);
-    });
-  });
+        return new Promise((res) => res(a * a))
+      }
+      const result = await _.map<number, number>(numbers, asyncSquare)
+      assert.deepEqual(result, [1, 4, 9, 16])
+    })
+  })
 
   describe("reduce/asyncReduceV2 function", () => {
-    const numbers = [0, 1, 2, 3, 4];
+    const numbers = [0, 1, 2, 3, 4]
     const reducer = async (a: number, b: number): Promise<number> => {
-      return new Promise((res) => res(a + b));
-    };
+      return new Promise((res) => res(a + b))
+    }
 
     test("calls asyncReduce", async () => {
-      const result = await _.reduce<number, number>(numbers, reducer, 0);
-      assert.equal(result, 10);
-    });
+      const result = await _.reduce<number, number>(numbers, reducer, 0)
+      assert.equal(result, 10)
+    })
     test("uses first item in array when no init provided", async () => {
-      const result = await _.reduce(numbers, reducer);
-      assert.equal(result, 10);
-    });
+      const result = await _.reduce(numbers, reducer)
+      assert.equal(result, 10)
+    })
     test("throws on no init value and an empty array", async () => {
       try {
-        await _.reduce([], reducer);
+        await _.reduce([], reducer)
       } catch (err) {
-        assert.isNotNull(err);
-        return;
+        assert.isNotNull(err)
+        return
       }
-      assert.fail("Expected error to be thrown");
-    });
+      assert.fail("Expected error to be thrown")
+    })
     test("throws on no init value and a null array input", async () => {
       try {
-        await _.reduce(null, reducer);
+        await _.reduce(null, reducer)
       } catch (err) {
-        assert.isNotNull(err);
-        return;
+        assert.isNotNull(err)
+        return
       }
-      assert.fail("Expected error to be thrown");
-    });
-  });
+      assert.fail("Expected error to be thrown")
+    })
+  })
 
   describe("map/asyncMapV2 function", () => {
     test("calls asyncMap", async () => {
-      const numbers = [1, 2, 3, 4];
-      const asyncSquare = async (a: number): Promise<number> => {
-        return new Promise((res) => res(a * a));
-      };
-      const result = await _.map<number, number>(numbers, asyncSquare);
-      assert.deepEqual(result, [1, 4, 9, 16]);
-    });
-  });
+      const numbers = [1, 2, 3, 4]
+	  let num = -1;
+      const asyncSquare = async (a: number, index: number): Promise<number> => {
+		num = index;
+        return new Promise((res) => res(a * a))
+      }
+      const result = await _.map<number, number>(numbers, asyncSquare)
+      assert.deepEqual(result, [1, 4, 9, 16])
+
+	  if(num === -1 || num !== numbers.length - 1) {
+		assert.fail("Expected asyncMap to be called with the correct index");
+	  }
```

Comment:
```suggestion
     const numbers = [1, 2, 3, 4]
      const indexes: number[] = []
      const asyncSquare = async (a: number, index: number): Promise<number> => {
        indexes.push(index)
        return new Promise((res) => res(a * a))
      }
      const result = await _.map<number, number>(numbers, asyncSquare)
      assert.deepEqual(result, [1, 4, 9, 16])
      assert.deepEqual(indexes, [0, 1, 2, 3])
```
A bit simpler test of the indexes, it also validates the index start and stop where we want. Looking at the code I wasn't sure if it would have an off-by-one issue because of the `i++` (I don't use i++ like ever so I don't know it's magic 😅 )